<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Longitudinal Analysis of Cancer Evolution (LACE) • LACE</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Longitudinal Analysis of Cancer Evolution (LACE)">
<meta property="og:description" content="
    LACE is an algorithmic framework that processes single-cell somatic mutation profiles from cancer samples collected at different time points and in distinct experimental settings, to produce longitudinal models of cancer evolution. The approach solves a Boolean Matrix Factorization problem with phylogenetic constraints, by maximizing a weighed likelihood function computed on multiple time points. ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">LACE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BIMIB-DISCo/LACE/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="longitudinal-analysis-of-cancer-evolution-lace" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#longitudinal-analysis-of-cancer-evolution-lace" class="anchor"></a>Longitudinal Analysis of Cancer Evolution (LACE)</h1></div>

<p>================</p>
<p><strong>OVERVIEW</strong></p>
<p>In this repository we provide an R implementation of <em>LACE</em>.</p>
<p><em>LACE</em> is an algorithmic framework that processes single-cell somatic mutation profiles from cancer samples collected at different time points and in distinct experimental settings, to produce longitudinal models of cancer evolution. The approach solves a Boolean Matrix Factorization problem with phylogenetic constraints, by maximizing a weighted likelihood function computed on multiple time points.</p>
<p><strong>INSTALLING LACE R IMPLEMENTATION</strong></p>
<p>The R version of <em>LACE</em> can be installed from Github. To do so, we need to install the R packages <em>LACE</em> depends on and the devtools package.</p>
<p>First we run an R session and we execute the following commands.</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="co"># run this commands only if the following R packages are not already installed</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"devtools"</span>, dependencies = <span class="fl">TRUE</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"igraph"</span>, dependencies = <span class="fl">TRUE</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"RColorBrewer"</span>, dependencies = <span class="fl">TRUE</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"Rfast"</span>, dependencies = <span class="fl">TRUE</span>)
</pre></div>
<p>Now we can install and run <em>LACE</em> as follows:</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="co"># install CIMLR from Github</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"devtools"</span>)
<span class="fu">install_github</span>(<span class="st">"BIMIB-DISCo/LACE"</span>, ref = <span class="st">"master"</span>)

<span class="co"># load LACE library</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st"><a href="https://github.com/BIMIB-DISCo/LACE">"LACE"</a></span>)
</pre></div>
<p><strong>RUNNING LACE R IMPLEMENTATION</strong></p>
<p>We now present an example of longitudinal analysis of cancer evolution with <em>LACE</em> using single-cell data obtained from Rambow, Florian, et al. “Toward minimal residual disease-directed therapy in melanoma.” Cell 174.4 (2018): 843-855.</p>
<p>As a first step, we load single-cell data for a skin melanoma. The data comprises point mutations for four time points: (1) before treatment, (2) 4 days treatment, (3) 28 days treatment and finally (4) 57 days treatment.</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st"><a href="https://github.com/BIMIB-DISCo/LACE">"LACE"</a></span>)
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="kw">longitudinal_sc_variants</span>)
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="kw">longitudinal_sc_variants</span>)

<span class="co">## [1] "T1_before_treatment"  "T2_4_days_treatment"  "T3_28_days_treatment" "T4_57_days_treatment"</span>
</pre></div>
<p>We setup the main parameter in oder to perform the inference. First of all, as the four data points may potentially provide sequencing for an unbalanced number of cells, we weight each time point proportionally to the sample sizes as follow. We refer to the paper for details.</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="kw">lik_weights</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.2308772</span>,<span class="fl">0.2554386</span>,<span class="fl">0.2701754</span>,<span class="fl">0.2435088</span>)
</pre></div>
<p>The second main parameter to be defined as input is represented by the false positive and false negative error rates, i.e., alpha and beta. We can specify a different rate per time point as a list of rates. When multiple set of rates are provided, <em>LACE</em> performs a grid search in order to estimate the best set of error rates.</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="kw">alpha</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>()
<span class="kw">alpha</span>[[<span class="fl">1</span>]] <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.02</span>,<span class="fl">0.01</span>,<span class="fl">0.01</span>,<span class="fl">0.01</span>)
<span class="kw">alpha</span>[[<span class="fl">2</span>]] <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.10</span>,<span class="fl">0.05</span>,<span class="fl">0.05</span>,<span class="fl">0.05</span>)
<span class="kw">beta</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>()
<span class="kw">beta</span>[[<span class="fl">1</span>]] <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.10</span>,<span class="fl">0.05</span>,<span class="fl">0.05</span>,<span class="fl">0.05</span>)
<span class="kw">beta</span>[[<span class="fl">2</span>]] <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.10</span>,<span class="fl">0.05</span>,<span class="fl">0.05</span>,<span class="fl">0.05</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw">alpha</span>)

<span class="co">## [[1]]</span>
<span class="co">## [1] 0.02 0.01 0.01 0.01</span>
<span class="co">##</span>
<span class="co">## [[2]]</span>
<span class="co">## [1] 0.10 0.05 0.05 0.05</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw">beta</span>)

<span class="co">## [[1]]</span>
<span class="co">## [1] 0.10 0.05 0.05 0.05</span>
<span class="co">##</span>
<span class="co">## [[2]]</span>
<span class="co">## [1] 0.10 0.05 0.05 0.05</span>
</pre></div>
<p>We can now perform the inference as follow.</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="kw">inference</span> <span class="op">=</span> <span class="fu"><a href="reference/LACE.html">LACE</a></span>(D = <span class="kw">longitudinal_sc_variants</span>, 
    lik_w = <span class="kw">lik_weights</span>, 
    alpha = <span class="kw">alpha</span>, 
    beta = <span class="kw">beta</span>, 
    keep_equivalent = <span class="fl">FALSE</span>, 
    num_rs = <span class="fl">5</span>, 
    num_iter = <span class="fl">10</span>, 
    n_try_bs = <span class="fl">5</span>, 
    num_processes = <span class="fl">NA</span>, 
    seed = <span class="fl">12345</span>, 
    verbose = <span class="fl">FALSE</span>)
</pre></div>
<p>We notice that the inference resulting on the command above should be considered only as an example; the parameters num_rs, num_iter and n_try_bs representing the number of steps perfomed during the inference are downscaled to reduce execution time. We refer to the Manual for discussion on default values. We provide within the package results of inferences performed with correct parameters as RData.</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="kw">inference</span>)
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="kw">inference</span>))

<span class="co">## [1] "B" "C" "corrected_genotypes"</span>
<span class="co">## [4] "clones_prevalence" "relative_likelihoods" "joint_likelihood"</span>
<span class="co">## [7] "clones_summary" "equivalent_solutions" "error_rates"</span>
</pre></div>
<p><em>LACE</em> returns a list of nine elements as results. Namely, B and C provide respectively the maximum likelihood longitudinal tree and cells attachments; corrected_genotypes the corrected genotypes, clones_prevalence, the estimated prevalence of any observed clone; relative_likelihoods and joint_likelihood the estimated likelihoods for each time point and the weighted likelihood; clones_summary provide a summary of association of mutations to clones. In equivalent_solutions, solutions (B and C) with likelihood equivalent to the best solution are returned; notice that in the example we disabled this feature by setting equivalent_solutions parameter to FALSE. Finally, error rates provide the best error rates (alpha and beta) as estimated by the grid search.</p>
<p>We can plot the inferred model using the function longitudinal.tree.plot.</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="kw">clone_labels</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"ARPC2"</span>,<span class="st">"PRAME"</span>,<span class="st">"HNRNPC"</span>,<span class="st">"COL1A2"</span>,<span class="st">"RPL5"</span>,<span class="st">"CCT8"</span>)
<span class="kw">longitudinal.tree</span> <span class="op">=</span> <span class="fu"><a href="reference/longitudinal.tree.plot.html">longitudinal.tree.plot</a></span>(inference = <span class="kw">inference</span>, 
                                           labels = <span class="st">"clones"</span>, 
                                           clone_labels = <span class="kw">clone_labels</span>, 
                                           legend_position = <span class="st">"topleft"</span>)
</pre></div>
<p><img src="LACE-Ex.jpg" alt="Image"></p>
<p><strong>DEBUG</strong></p>
<p>Please feel free to contact us if you have problems running our tool at <a href="mailto:daniele.ramazzotti1@gmail.com" class="email">daniele.ramazzotti1@gmail.com</a> or <a href="mailto:d.maspero@campus.unimib.it" class="email">d.maspero@campus.unimib.it</a>.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from BIOC at <br><a href="https://www.bioconductor.org/packages/LACE">https://​www.bioconductor.org/​packages/​LACE</a>
</li>
<li>Browse source code at <br><a href="https://github.com/BIMIB-DISCo/LACE/">https://​github.com/​BIMIB-DISCo/​LACE/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/BIMIB-DISCo/LACE">https://​github.com/​BIMIB-DISCo/​LACE</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Daniele Ramazzotti <br><small class="roles"> Author </small>  </li>
<li>Fabrizio Angaroni <br><small class="roles"> Author </small>  </li>
<li>Davide Maspero <br><small class="roles"> Maintainer, author </small>  </li>
<li>Alex Graudenzi <br><small class="roles"> Author </small>  </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/BIMIB-DISCo/LACE/actions?query=workflow%3Acheck-master"><img src="https://github.com/BIMIB-DISCo/LACE/workflows/check-master/badge.svg" alt="Actions Status"></a></li>
<li><a href="https://github.com/BIMIB-DISCo/LACE/actions?query=workflow%3Acheck-development"><img src="https://github.com/BIMIB-DISCo/LACE/workflows/check-development/badge.svg" alt="Actions Status"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Daniele Ramazzotti, Fabrizio Angaroni, Davide Maspero, Alex Graudenzi.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
