<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Longitudinal Analysis of Cancer Evolution (LACE) • LACE</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Longitudinal Analysis of Cancer Evolution (LACE)">
<meta property="og:description" content="
    LACE is an algorithmic framework that processes single-cell somatic mutation profiles from cancer samples collected at different time points and in distinct experimental settings, to produce longitudinal models of cancer evolution. The approach solves a Boolean Matrix Factorization problem with phylogenetic constraints, by maximizing a weighed likelihood function computed on multiple time points.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">LACE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/1_introduction.html">Introduction</a>
    </li>
    <li>
      <a href="articles/2_running_LACE.html">Running LACE</a>
    </li>
    <li>
      <a href="articles/3_LACE_interface.html">LACE 2.0</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BIMIB-DISCo/LACE/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="lace-20-an-interactive-r-tool-for-the-inference-and-visualization-of-longitudinal-cancer-evolution">LACE 2.0: an interactive R tool for the inference and visualization of longitudinal cancer evolution<a class="anchor" aria-label="anchor" href="#lace-20-an-interactive-r-tool-for-the-inference-and-visualization-of-longitudinal-cancer-evolution"></a>
</h1></div>
<p><a href="https://github.com/BIMIB-DISCo/LACE/actions?query=workflow%3Acheck-master" class="external-link"><img src="https://github.com/BIMIB-DISCo/LACE/workflows/check-master/badge.svg" alt="Actions Status"></a> <a href="https://github.com/BIMIB-DISCo/LACE/actions?query=workflow%3Acheck-development" class="external-link"><img src="https://github.com/BIMIB-DISCo/LACE/workflows/check-development/badge.svg" alt="Actions Status"></a> <a href="http://bioconductor.org/packages/release/bioc/html/LACE.html#since" class="external-link"><img border="0" src="http://bioconductor.org/shields/years-in-bioc/LACE.svg" title="How long since the package was first in a released Bioconductor version (or is it in devel only)."></a></p>
<p>LACE 2.0 is a new release of the LACE R Bioconductor package. LACE 2.0 is capable of performing clonal evolution analyses for single-cell sequencing data including longitudinal experiments. LACE 2.0 allows to annotate variants and retrieve the relevant mutations interactively based on user-defined filtering criteria; it infers the maximum likelihood clonal tree, cell matrix attachment and false positive/negative rates using boolean matrix factorization. Furthermore, it generates the longitudinal clonal tree. Finally, LACE 2.0 allows to investigate cancer clonal evolution under different experimental conditions and the occurrence of single mutations which can be queried via <em>ensembl</em> database.</p>
<p>A detailed vignette on LACE 2.0 is provided here: <a href="https://bimib-disco.github.io/LACE/articles/3_LACE_interface.html" class="external-link uri">https://bimib-disco.github.io/LACE/articles/3_LACE_interface.html</a></p>
<div class="section level2">
<h2 id="citation">CITATION<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>When using our tool, please cite: Ramazzotti, Daniele, et al. “LACE: Inference of cancer evolution models from longitudinal single-cell sequencing data.” Journal of Computational Science 58 (2022): 101523.</p>
</div>
<div class="section level2">
<h2 id="installation-of-lace-20-r-package">Installation of LACE 2.0 R package<a class="anchor" aria-label="anchor" href="#installation-of-lace-20-r-package"></a>
</h2>
<p>The package is available on GitHub and Bioconductor. LACE 2.0 requires R &gt; 4.1.0.</p>
<p>To install from Bioconductor, please run:</p>
<pre><code><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"LACE"</span><span class="op">)</span></span></code></pre>
<p>To install LACE 2.0 from GitHub run:</p>
<pre><code><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">"devtools"</a></span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">"devtools"</a></span><span class="op">)</span></span>
<span><span class="fu">install_github</span><span class="op">(</span><span class="st">"BIMIB-DISCo/LACE"</span>, ref <span class="op">=</span> <span class="st">"development"</span>, dependencies <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
<p>LACE 2.0 uses <em>Annovar</em> and <em>Samtools suite</em> as back-ends for variant calling annotation and depth computation, respectively. Please refer to the next section to install them.</p>
</div>
<div class="section level2">
<h2 id="installation-of-other-required-softwares">Installation of other required softwares<a class="anchor" aria-label="anchor" href="#installation-of-other-required-softwares"></a>
</h2>
<p><em>Annovar</em> is a widely used variant calling software freely available upon registration to their website at <a href="https://annovar.openbioinformatics.org/en/latest/" class="external-link uri">https://annovar.openbioinformatics.org/en/latest/</a>. The package contains <em>Perl</em> scripts and variant calling annotation reference databases for the human species. For other databases, please refer to their website. If the scripts are installed in binary search path, then LACE 2.0 will detect them automatically.</p>
<p><em>Perl</em> (<a href="https://www.perl.org/" class="external-link uri">https://www.perl.org/</a>) is required to run <em>Annovar</em>.</p>
<p><em>Samtools suite</em> is a standard set of tools and libraries to handle SAM/BAM/BED file format and perform a variety of common operations on sequencing data. It is freely available at <a href="http://www.htslib.org/" class="external-link uri">http://www.htslib.org/</a> and <a href="https://github.com/samtools/htslib" class="external-link uri">https://github.com/samtools/htslib</a>. To install <em>Samtools</em> follow the instructions in their website.</p>
<div class="section level4">
<h4 id="for-windows-users-we-suggest-the-following-guidelines">For <em>Windows</em> users, we suggest the following guidelines:<a class="anchor" aria-label="anchor" href="#for-windows-users-we-suggest-the-following-guidelines"></a>
</h4>
<ul>
<li><p>Download <em>MSYS2</em> or <em>WSL</em></p></li>
<li><p>Download the <em>Samtools</em> source files from <a href="http://www.htslib.org/" class="external-link uri">http://www.htslib.org/</a></p></li>
<li><p>The field <code>db_home</code>, in the <em>Samtools</em> source file <code>etc/nsswitch.conf</code>, should be changed to <code>windows</code> such that: <code>db_home: windows</code></p></li>
<li><p>Install <em>MSYS2</em>/<em>WSL</em> (it is preferably to have <em>MSYS2</em> in the “C:” path), and install the packages required by Samtools as stated in the INSTALL documentation file within the Samtool source folder</p></li>
<li><p>Inside a <em>MSYS2</em>/<em>WSL</em> shell, add the following directories to the variable PATH using the command: <code>export PATH="/mingw64/bin/:/mingw64/:$PATH"</code></p></li>
<li><p>From the above <em>MSYS2</em>/<em>WSL</em> shell, follow the Samtools documentation to build and install the software</p></li>
<li><p>Change the Windows <code>PATH</code> variable in the System variables and add the following paths: <code>C:\msys64\usr\bin          C:\msys64\usr          C:\msys64\mingw64\bin          C:\msys64\mingw64</code></p></li>
<li><p>We remind that <em>Annovar</em> is a <em>Perl</em> script, and <code>.pl</code> files need be associated to <em>Perl</em> executable.</p></li>
</ul>
<p>Eventually, from the <em>Windows</em> command prompt, users should be able to start <em>Samtools</em> using the command <code>samtools</code> and directly execute <em>Perl</em> scripts by calling their filenames.</p>
</div>
</div>
<div class="section level2">
<h2 id="run-lace-20">Run LACE 2.0<a class="anchor" aria-label="anchor" href="#run-lace-20"></a>
</h2>
<p>To start LACE 2.0 user interface run:</p>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/BIMIB-DISCo/LACE" class="external-link">LACE</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/LACEview.html">LACEview</a></span><span class="op">(</span><span class="op">)</span></span></code></pre>
<div class="figure">
<img src="https://github.com/BIMIB-DISCo/LACE/blob/master/vignettes/resources/Display_tab.png?raw=true" alt=""><p class="caption">Picture</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://www.bioconductor.org/packages/LACE" class="external-link">View on Bioconductor</a></li>
<li><a href="https://github.com/BIMIB-DISCo/LACE/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/BIMIB-DISCo/LACE" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing LACE</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Daniele Ramazzotti <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-6087-2666" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Fabrizio Angaroni <br><small class="roles"> Author </small>  </li>
<li>Davide Maspero <br><small class="roles"> Maintainer, author </small>  </li>
<li>Alex Graudenzi <br><small class="roles"> Author </small>  </li>
<li>Luca De Sano <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-9618-3774" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Gianluca Ascolani <br><small class="roles"> Author </small>  </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Daniele Ramazzotti, Fabrizio Angaroni, Davide Maspero, Alex Graudenzi, Luca De Sano, Gianluca Ascolani.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
