<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>LACE — LACE • LACE</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="LACE — LACE"><meta property="og:description" content="Perform the inference of the maximum likelihood clonal tree from longitudinal data."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">LACE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.8.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/v1_introduction.html">Introduction</a>
    </li>
    <li>
      <a href="../articles/v2_running_LACE.html">Running LACE</a>
    </li>
    <li>
      <a href="../articles/v3_LACE_interface.html">LACE-interface</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/BIMIB-DISCo/LACE/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>LACE</h1>
    <small class="dont-index">Source: <a href="https://github.com/BIMIB-DISCo/LACE/blob/master/R/frontend.R" class="external-link"><code>R/frontend.R</code></a></small>
    <div class="hidden name"><code>LACE.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Perform the inference of the maximum likelihood clonal tree from longitudinal data.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">LACE</span><span class="op">(</span></span>
<span>  <span class="va">D</span>,</span>
<span>  lik_w <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  alpha <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  beta <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  initialization <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  random_tree <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  keep_equivalent <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  check_indistinguishable <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  num_rs <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  num_iter <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  n_try_bs <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  learning_rate <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  marginalize <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  error_move <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  num_processes <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  seed <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  log_file <span class="op">=</span> <span class="st">""</span>,</span>
<span>  show <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-d">D<a class="anchor" aria-label="anchor" href="#arg-d"></a></dt>
<dd><p>Mutation data from multiple experiments for a list of driver genes. It can be either a list with a data matrix per time point or a SummarizedExperiment object.
In this latter, the object must contain two fields: assays and colData. Assays stores one unique data matrix pooling all single cells observed at each time point and colData stores a vector of labels reporting the time point when each single cell was sequenced.
Ordering of cells in assays field and colData field must be the same.</p></dd>


<dt id="arg-lik-w">lik_w<a class="anchor" aria-label="anchor" href="#arg-lik-w"></a></dt>
<dd><p>Weight for each data point. If not provided, weights to correct for sample sizes are used.</p></dd>


<dt id="arg-alpha">alpha<a class="anchor" aria-label="anchor" href="#arg-alpha"></a></dt>
<dd><p>False positive error rate provided as list of elements; if a vector of alpha (and beta) is provided, the inference is performed for multiple values and the solution at
maximum-likelihood is returned.</p></dd>


<dt id="arg-beta">beta<a class="anchor" aria-label="anchor" href="#arg-beta"></a></dt>
<dd><p>False negative error rate provided as list of elements; if a vector of beta (and alpha) is provided, the inference is performed for multiple values and the solution at
maximum-likelihood is returned.</p></dd>


<dt id="arg-initialization">initialization<a class="anchor" aria-label="anchor" href="#arg-initialization"></a></dt>
<dd><p>Binary matrix representing a perfect philogeny clonal tree; clones are rows and mutations are columns.
This parameter overrides "random_tree".</p></dd>


<dt id="arg-random-tree">random_tree<a class="anchor" aria-label="anchor" href="#arg-random-tree"></a></dt>
<dd><p>Boolean. Shall I start MCMC search from a random tree? If FALSE (default) and initialization is NULL, search
is started from a TRaIT tree (BMC Bioinformatics . 2019 Apr 25;20(1):210.  doi: 10.1186/s12859-019-2795-4).</p></dd>


<dt id="arg-keep-equivalent">keep_equivalent<a class="anchor" aria-label="anchor" href="#arg-keep-equivalent"></a></dt>
<dd><p>Boolean. Shall I return results (B and C) at equivalent likelihood with the best returned solution?</p></dd>


<dt id="arg-check-indistinguishable">check_indistinguishable<a class="anchor" aria-label="anchor" href="#arg-check-indistinguishable"></a></dt>
<dd><p>Boolean. Shall I remove any indistinguishable event from input data prior inference?</p></dd>


<dt id="arg-num-rs">num_rs<a class="anchor" aria-label="anchor" href="#arg-num-rs"></a></dt>
<dd><p>Number of restarts during mcmc inference.</p></dd>


<dt id="arg-num-iter">num_iter<a class="anchor" aria-label="anchor" href="#arg-num-iter"></a></dt>
<dd><p>Maximum number of mcmc steps to be performed during the inference.</p></dd>


<dt id="arg-n-try-bs">n_try_bs<a class="anchor" aria-label="anchor" href="#arg-n-try-bs"></a></dt>
<dd><p>Number of steps without change in likelihood of best solution after which to stop the mcmc.</p></dd>


<dt id="arg-learning-rate">learning_rate<a class="anchor" aria-label="anchor" href="#arg-learning-rate"></a></dt>
<dd><p>Parameter to tune the probability of accepting solutions at lower values during mcmc. Value of learning_rate = 1 (default), set a
probability proportional to the difference in likelihood; values of learning_rate greater than 1 inclease the chance of accepting solutions at lower likelihood
during mcmc while values lower than 1 decrease such probability.</p></dd>


<dt id="arg-marginalize">marginalize<a class="anchor" aria-label="anchor" href="#arg-marginalize"></a></dt>
<dd><p>Boolean. Shall I marginalize C when computing likelihood?</p></dd>


<dt id="arg-error-move">error_move<a class="anchor" aria-label="anchor" href="#arg-error-move"></a></dt>
<dd><p>Boolean. Shall I include estimation of error rates in the MCMC moves?</p></dd>


<dt id="arg-num-processes">num_processes<a class="anchor" aria-label="anchor" href="#arg-num-processes"></a></dt>
<dd><p>Number of processes to be used during parallel execution. To execute in single process mode,
this parameter needs to be set to either NA or NULL.</p></dd>


<dt id="arg-seed">seed<a class="anchor" aria-label="anchor" href="#arg-seed"></a></dt>
<dd><p>Seed for reproducibility.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Boolean. Shall I print to screen information messages during the execution?</p></dd>


<dt id="arg-log-file">log_file<a class="anchor" aria-label="anchor" href="#arg-log-file"></a></dt>
<dd><p>log file where to print outputs when using parallel. If parallel execution is disabled, this parameter is ignored.</p></dd>


<dt id="arg-show">show<a class="anchor" aria-label="anchor" href="#arg-show"></a></dt>
<dd><p>Boolean. Show the interactive interface to explore the output.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A list of 9 elements: B, C, clones_prevalence, relative_likelihoods, joint_likelihood, clones_summary and error_rates. Here, B returns the maximum likelihood longitudinal
clonal tree, C the attachment of cells to clones, corrected_genotypes the corrected genotypes and clones_prevalence clones' prevalence; relative_likelihoods and joint_likelihood are respectively
the likelihood of the solutions at each individual time points and the joint likelihood; clones_summary provide a summary of association of mutations to clones. In equivalent_solutions,
solutions (B and C) with likelihood equivalent to the best solution are returned. Finally error_rates provides the best values of alpha and beta among the considered ones.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">longitudinal_sc_variants</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">inference</span> <span class="op">=</span> <span class="fu">LACE</span><span class="op">(</span>D <span class="op">=</span> <span class="va">longitudinal_sc_variants</span>,</span></span>
<span class="r-in"><span>                 lik_w <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.2308772</span>,<span class="fl">0.2554386</span>,<span class="fl">0.2701754</span>,<span class="fl">0.2435088</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                 alpha <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.10</span>,<span class="fl">0.05</span>,<span class="fl">0.05</span>,<span class="fl">0.05</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                 beta <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.10</span>,<span class="fl">0.05</span>,<span class="fl">0.05</span>,<span class="fl">0.05</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                 keep_equivalent <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>                 num_rs <span class="op">=</span> <span class="fl">5</span>,</span></span>
<span class="r-in"><span>                 num_iter <span class="op">=</span> <span class="fl">10</span>,</span></span>
<span class="r-in"><span>                 n_try_bs <span class="op">=</span> <span class="fl">5</span>,</span></span>
<span class="r-in"><span>                 num_processes <span class="op">=</span> <span class="cn">NA</span>,</span></span>
<span class="r-in"><span>                 seed <span class="op">=</span> <span class="fl">12345</span>,</span></span>
<span class="r-in"><span>                 verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>                 show <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Daniele Ramazzotti, Fabrizio Angaroni, Davide Maspero, Alex Graudenzi, Luca De Sano, Gianluca Ascolani.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.9000.</p>
</div>

      </footer></div>






  </body></html>

