<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>LACE — LACE • LACE</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="LACE — LACE" />
<meta property="og:description" content="Perform inference of the maximum likelihood clonal tree from longitudinal data." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">LACE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/BIMIB-DISCo/LACE/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>LACE</h1>
    <small class="dont-index">Source: <a href='https://github.com/BIMIB-DISCo/LACE/blob/master/R/frontend.R'><code>R/frontend.R</code></a></small>
    <div class="hidden name"><code>LACE.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Perform inference of the maximum likelihood clonal tree from longitudinal data.</p>
    </div>

    <pre class="usage"><span class='fu'>LACE</span>(
  <span class='kw'>D</span>,
  lik_w = <span class='kw'>NULL</span>,
  alpha = <span class='kw'>NULL</span>,
  beta = <span class='kw'>NULL</span>,
  initialization = <span class='kw'>NULL</span>,
  keep_equivalent = <span class='fl'>TRUE</span>,
  check_indistinguishable = <span class='fl'>TRUE</span>,
  num_rs = <span class='fl'>50</span>,
  num_iter = <span class='fl'>10000</span>,
  n_try_bs = <span class='fl'>500</span>,
  learning_rate = <span class='fl'>1</span>,
  marginalize = <span class='fl'>FALSE</span>,
  num_processes = <span class='fl'>Inf</span>,
  seed = <span class='kw'>NULL</span>,
  verbose = <span class='fl'>TRUE</span>,
  log_file = <span class='st'>""</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>D</th>
      <td><p>Mutation data from multiple experiments for a list of driver genes. It can be either a list with a data matrix per time point or a SummarizedExperiment object. 
In this latter, the object must contain two fields: assays and colData. Assays stores one unique data matrix pooling all single cells observed at each time point and colData stores a vector of labels reporting the time point when each single cell was sequenced.
Ordering of cells in assays field and colData field must be the same.</p></td>
    </tr>
    <tr>
      <th>lik_w</th>
      <td><p>Weight for each data point. If not provided, weights to correct for sample sizes are used.</p></td>
    </tr>
    <tr>
      <th>alpha</th>
      <td><p>False positive error rate provided as list of elements; if a vector of alpha (and beta) is provided, the inference is performed for multiple values and the solution at 
maximum-likelihood is returned.</p></td>
    </tr>
    <tr>
      <th>beta</th>
      <td><p>False negative error rate provided as list of elements; if a vector of beta (and alpha) is provided, the inference is performed for multiple values and the solution at 
maximum-likelihood is returned.</p></td>
    </tr>
    <tr>
      <th>initialization</th>
      <td><p>Starting point of the mcmc; if not provided, a random starting point is used.</p></td>
    </tr>
    <tr>
      <th>keep_equivalent</th>
      <td><p>Boolean. Shall I return results (B and C) at equivalent likelihood with the best returned solution?</p></td>
    </tr>
    <tr>
      <th>check_indistinguishable</th>
      <td><p>Boolean. Shall I remove any indistinguishable event from input data prior inference?</p></td>
    </tr>
    <tr>
      <th>num_rs</th>
      <td><p>Number of restarts during mcmc inference.</p></td>
    </tr>
    <tr>
      <th>num_iter</th>
      <td><p>Maximum number of mcmc steps to be performed during the inference.</p></td>
    </tr>
    <tr>
      <th>n_try_bs</th>
      <td><p>Number of steps without change in likelihood of best solution after which to stop the mcmc.</p></td>
    </tr>
    <tr>
      <th>learning_rate</th>
      <td><p>Parameter to tune the probability of accepting solutions at lower values during mcmc. Value of learning_rate = 1 (default), set a 
probability proportional to the difference in likelihood; values of learning_rate greater than 1 inclease the chance of accepting solutions at lower likelihood 
during mcmc while values lower than 1 decrease such probability.</p></td>
    </tr>
    <tr>
      <th>marginalize</th>
      <td><p>Boolean. Shall I marginalize C when computing likelihood?</p></td>
    </tr>
    <tr>
      <th>num_processes</th>
      <td><p>Number of processes to be used during parallel execution. To execute in single process mode, 
this parameter needs to be set to either NA or NULL.</p></td>
    </tr>
    <tr>
      <th>seed</th>
      <td><p>Seed for reproducibility.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>Boolean. Shall I print to screen information messages during the execution?</p></td>
    </tr>
    <tr>
      <th>log_file</th>
      <td><p>log file where to print outputs when using parallel. If parallel execution is disabled, this parameter is ignored.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list of 9 elements: B, C, clones_prevalence, relative_likelihoods, joint_likelihood, clones_summary and error_rates. Here, B returns the maximum likelihood longitudinal 
clonal tree, C the attachment of cells to clones, corrected_genotypes the corrected genotypes and clones_prevalence clones' prevalence; relative_likelihoods and joint_likelihood are respectively 
the likelihood of the solutions at each individual time points and the joint likelihood; clones_summary provide a summary of association of mutations to clones. In equivalent_solutions, 
solutions (B and C) with likelihood equivalent to the best solution are returned. Finally error_rates provides the best values of alpha and beta among the considered ones.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='kw'>longitudinal_sc_variants</span>)
<span class='kw'>inference</span> <span class='op'>=</span> <span class='fu'>LACE</span>(D = <span class='kw'>longitudinal_sc_variants</span>,
                 lik_w = <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>0.2308772</span>,<span class='fl'>0.2554386</span>,<span class='fl'>0.2701754</span>,<span class='fl'>0.2435088</span>),
                 alpha = <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>0.10</span>,<span class='fl'>0.05</span>,<span class='fl'>0.05</span>,<span class='fl'>0.05</span>)),
                 beta = <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>0.10</span>,<span class='fl'>0.05</span>,<span class='fl'>0.05</span>,<span class='fl'>0.05</span>)),
                 keep_equivalent = <span class='fl'>FALSE</span>,
                 num_rs = <span class='fl'>5</span>,
                 num_iter = <span class='fl'>10</span>,
                 n_try_bs = <span class='fl'>5</span>,
                 num_processes = <span class='fl'>NA</span>,
                 seed = <span class='fl'>12345</span>,
                 verbose = <span class='fl'>FALSE</span>)
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Daniele Ramazzotti, Fabrizio Angaroni, Davide Maspero, Alex Graudenzi.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


