<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>LACE-interface • LACE</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="LACE-interface">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">LACE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.10.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/v1_introduction.html">Introduction</a>
    </li>
    <li>
      <a href="../articles/v2_running_LACE.html">Running LACE</a>
    </li>
    <li>
      <a href="../articles/v3_LACE_interface.html">LACE-interface</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BIMIB-DISCo/LACE/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>LACE-interface</h1>
                        <h4 data-toc-skip class="author">Daniele
Ramazzotti</h4>
                        <h4 data-toc-skip class="author">Fabrizio
Angaroni</h4>
                        <h4 data-toc-skip class="author">Davide
Maspero</h4>
                        <h4 data-toc-skip class="author">Alex
Graudenxi</h4>
                        <h4 data-toc-skip class="author">Luca De
Sano</h4>
                        <h4 data-toc-skip class="author">Gianluca
Ascolani</h4>
            
            <h4 data-toc-skip class="date">October 03, 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BIMIB-DISCo/LACE/blob/master/vignettes/v3_LACE_interface.Rmd" class="external-link"><code>vignettes/v3_LACE_interface.Rmd</code></a></small>
      <div class="hidden name"><code>v3_LACE_interface.Rmd</code></div>

    </div>

    
    
<p>LACE 2.0 is a new release of the LACE R package. LACE 2.0 is capable
of performing clonal evolution analyses for single-cell sequencing data
including longitudinal experiments. LACE 2.0 allows to annotate variants
and retrieve the relevant mutations interactively based on user-defined
filtering criteria; it infers the maximum likelihood clonal tree, cell
matrix attachments and false positive/negative rates using boolean
matrix factorization. Furthermore, LACE 2.0 allows to investigate cancer
clonal evolution under different experimental conditions and the
occurrence of single mutations which can be queried via <em>ensembl</em>
database.</p>
<div class="section level2">
<h2 id="installation-of-lace-2-0-r-package">Installation of LACE 2.0 R package<a class="anchor" aria-label="anchor" href="#installation-of-lace-2-0-r-package"></a>
</h2>
<p>The package is available on GitHub and Bioconductor. LACE 2.0
requires R &gt;= 4.2.0 and Bioconductor. To install Bioconductor
run:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"LACE"</span><span class="op">)</span></span></code></pre></div>
<p>LACE 2.0 uses <em>Annovar</em> and <em>Samtools suite</em> as
back-ends for variant calling annotation and depth computation,
respectively. Please refer to the next section to install them.</p>
</div>
<div class="section level2">
<h2 id="installation-of-other-required-softwares">Installation of other required softwares<a class="anchor" aria-label="anchor" href="#installation-of-other-required-softwares"></a>
</h2>
<p><em>Annovar</em> is a widely used variant calling software freely
available upon registration to their website at <a href="https://annovar.openbioinformatics.org/en/latest/" class="external-link uri">https://annovar.openbioinformatics.org/en/latest/</a>. The
package contains <em>Perl</em> scripts and variant calling annotation
reference databases for the human species. For other databases, please
refer to their website. If the scripts are installed in binary search
path, then LACE 2.0 will detect them automatically.</p>
<p><em>Perl</em> (<a href="https://www.perl.org/" class="external-link uri">https://www.perl.org/</a>) is required to run
<em>Annovar</em>.</p>
<p><em>Samtools suite</em> is a standard set of tools and libraries to
handle SAM/BAM/BED file format and perform a variety of common
operations on sequencing data. It is freely available at <a href="http://www.htslib.org/" class="external-link uri">http://www.htslib.org/</a> and
<a href="https://github.com/samtools/htslib" class="external-link uri">https://github.com/samtools/htslib</a>. To install
<em>Samtools</em> follow the instructions in their website.</p>
<div class="section level4">
<h4 id="for-windows-users-we-suggest-the-following-guidelines">For <em>Windows</em> users, we suggest the following
guidelines:<a class="anchor" aria-label="anchor" href="#for-windows-users-we-suggest-the-following-guidelines"></a>
</h4>
<ul>
<li><p>Download <em>MSYS2</em> or <em>WSL</em></p></li>
<li><p>Download the <em>Samtools</em> source files from <a href="http://www.htslib.org/" class="external-link uri">http://www.htslib.org/</a></p></li>
<li><p>The field <code>db_home</code>, in the <em>Samtools</em> source
file <code>etc/nsswitch.conf</code>, should be changed to
<code>windows</code> such that: <code>db_home: windows</code></p></li>
<li><p>Install <em>MSYS2</em>/<em>WSL</em> (it is preferably to have
<em>MSYS2</em> in the “C:” path), and install the packages required by
Samtools as stated in the INSTALL documentation file within the Samtool
source folder</p></li>
<li><p>Inside a <em>MSYS2</em>/<em>WSL</em> shell, add the following
directories to the variable PATH using the command:
<code>export PATH="/mingw64/bin/:/mingw64/:$PATH"</code></p></li>
<li><p>From the above <em>MSYS2</em>/<em>WSL</em> shell, follow the
Samtools documentation to build and install the software</p></li>
<li><p>Change the Windows <code>PATH</code> variable in the System
variables and add the following paths: <code>C:\msys64\usr\bin</code>,
<code>C:\msys64\usr</code>, <code>C:\msys64\mingw64\bin</code>,
<code>C:\msys64\mingw64</code></p></li>
<li><p>We remind that <em>Annovar</em> is a <em>Perl</em> script, and
<code>.pl</code> files need be associated to <em>Perl</em>
executable.<br>
Eventually, from the <em>Windows</em> command prompt, users should be
able to start <em>Samtools</em> using the command <code>samtools</code>
and directly execute <em>Perl</em> scripts by calling their
filenames.</p></li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="running-lace-2-0">Running LACE 2.0<a class="anchor" aria-label="anchor" href="#running-lace-2-0"></a>
</h2>
<p>To start LACE 2.0 user interface run:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/BIMIB-DISCo/LACE" class="external-link">LACE</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/LACEview.html">LACEview</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="using-lace-2-0">Using LACE 2.0<a class="anchor" aria-label="anchor" href="#using-lace-2-0"></a>
</h2>
<p>LACE 2.0 has been thought to be used on single cell sequencing data
for which it is available variant calling data in standard VCF format
and binary aligned data in standard BAM format.</p>
<p>The user is provided with an interface to initiate a project and to
set filter thresholds after which annotation of variants, filtering of
data and depth at variant sites are retrieved. Both annotation and depth
derivation are computationally expensive steps. LACE 2.0 reduces
possible re-computation by detecting parameter variations of the user
interface and by comparison of the timestamps of interface state, inputs
and outputs. Intermediary and final data are stored in the designated
folders.</p>
<p>The operation is followed by the possibility for the user to select
variants which are drivers of the understudied biological problem.</p>
<p>At this point, the inferential step is executed so that the most
likelihood longitudinal clonal tree and clonal prevalences are retrieved
together with the best set of false positive/negative rates among those
provided by the user.</p>
<p>The results are displayed via an interactive interface.</p>
</div>
<div class="section level2">
<h2 id="interface">Interface<a class="anchor" aria-label="anchor" href="#interface"></a>
</h2>
<p>The interface is divided in two parts interleaved by variant
selection and inferential computation parts: the processing interface
and the results interface.</p>
<div class="section level3">
<h3 id="project-creation">Project creation<a class="anchor" aria-label="anchor" href="#project-creation"></a>
</h3>
<p>To begin the clonal analyses the user needs to create a project by
choosing a folder and a meaningful name for the project (field 1.1 and
1.2).</p>
<p><img src="resources%2FProject_tab.png" style="width:95.0%"></p>
<p>If the project path does not contain former LACE projects, it asks
the user to create a new one (“Create project”) and all intermediate and
final results will be stored in structured sub-folders of the chosen
project path.</p>
<p>LACE 2.0 automatically recognizes if the selected path is a LACE
project and proposes the user to load it (<img src="resources%2Fload_project.png" style="width:10.0%">)</p>
</div>
<div class="section level3">
<h3 id="sidebar-and-demos">Sidebar and Demos<a class="anchor" aria-label="anchor" href="#sidebar-and-demos"></a>
</h3>
<p>The sidebar is divided in Projects and Current Project. Projects
contain a set of recently opened projects which can be reloaded by
selecting the respective project name.</p>
<p><img src="resources%2FProjects_menu.png" style="width:30.0%"></p>
<p>Under Projects, there is the sub-menu Demos. LACE 2.0 provides the
user with two demos:</p>
<ul>
<li>Melanoma dataset [Rambow, Florian, et al. “Toward minimal residual
disease-directed therapy in melanoma.” Cell 174.4 (2018): 843-855.]
containing single cell variant annotations and depths for a total of 674
cells sampled at 4 different time points:
<ol style="list-style-type: decimal">
<li>sample before treatment
<ul>
<li>BRAF inhibitor treatment</li>
</ul>
</li>
<li>sample after 4 days</li>
<li>sample after 28 days</li>
<li>sample after 57 days</li>
</ol>
</li>
<li>Small dataset containing only 3 cells per 4 sampling time
points.</li>
</ul>
<p>Current Project allows the user to save and load the parameters of
the whole project or of the active tab.</p>
</div>
<div class="section level3">
<h3 id="processing-interface">Processing interface<a class="anchor" aria-label="anchor" href="#processing-interface"></a>
</h3>
<div class="section level4">
<h4 id="single-cell-metadata">Single Cell Metadata<a class="anchor" aria-label="anchor" href="#single-cell-metadata"></a>
</h4>
<p>In order to perform clonal analyses it is required to provide some
information about the experiment such as the cell IDs which are used to
retrieve the VCF/BAM files and the sample name each cell belongs to.
These metadata are generally included with the experimental sequencing
data as part of the library preparation. LACE 2.0 accepts tsv/csv
tabular formats with headers and rds file format containing a data.frame
(field 2.1). Metadata file can include more information relative to the
experimental setup and the platform used.</p>
<p><img src="resources%2Ffigure2c.png" style="width:95.0%"></p>
<p>The user should identify and select the ID (field 2.2) and sampling
columns (field 2.3). The user can also reorder the sampling names so to
recreate the chronological order of longitudinal events of the
experiment, or to explore different orders (field 2.4).</p>
</div>
<div class="section level4">
<h4 id="annotations">Annotations<a class="anchor" aria-label="anchor" href="#annotations"></a>
</h4>
<p>Annotations of variant calling data is performed using
<em>Annovar</em> as back-end. Each mutation of each cell is annotated
based on the annotation database used. <em>Annovar</em> provides a
database for the human species useful to tag cancer mutations and
whenever possible their functional effect.</p>
<p><img src="resources%2FAnnotation_tab.png" style="width:95.0%"></p>
<p>If <em>Annovar</em>’s <em>Perl</em> scripts are available in the
standard binary paths, they are detected and the <em>Annovar</em> folder
field is autocompleted; otherwise, the user should provide the folder
containing the <em>Perl</em> scripts. The user should also provide the
database to use for the annotation and the folder containing the variant
calling files (VCF).</p>
</div>
<div class="section level4">
<h4 id="variant-filtering">Variant filtering<a class="anchor" aria-label="anchor" href="#variant-filtering"></a>
</h4>
<p>All types of single cell sequencing data is characterized by various
sources of noise which depends on the technology used. Detected mutation
might be characterized by low quality score or low statistical power.
Some mutations can be neglected, while others might cause relevant
effects, especially on exonic regions where variations can result in
changes of the translational process and modify their functional
form.</p>
<p>In order to avoid small and possibly spurious fluctuations on
sequencing data, variants can be filtered if they have low supporting
evidences. The user can set the minimum values for:</p>
<ol style="list-style-type: decimal">
<li>the number of reads supporting the alternative alleles in a cell
(field 3.1),</li>
<li>the frequency of the minor allele for each referenced SNP included
in a default global population (field 3.2),</li>
<li>the cell frequency per sample showing the mutation at the same site
(field 3.3).</li>
</ol>
<p><img src="resources%2Ffigure3c.png" style="width:95.0%"></p>
<p>Furthermore, the user can choose which functional exonic variation
should be considered or neglected (field 3.4). For example, there are
cases in which unknown and synonymous mutations in exonic regions are
disregarded because their effect cannot be explicitly related to the
experimental condition. All possible variant functions and their
explanations are resumed in the following table:</p>
<table class="table">
<colgroup>
<col width="30%">
<col width="69%">
</colgroup>
<thead><tr class="header">
<th>Annotation (+)</th>
<th>Explanation</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>frameshift insertion</td>
<td>an insertion of one or more nucleotides that cause frameshift
changes in protein coding sequence</td>
</tr>
<tr class="even">
<td>frameshift deletion</td>
<td>a deletion of one or more nucleotides that cause frameshift changes
in protein coding sequence</td>
</tr>
<tr class="odd">
<td>frameshift block substitution</td>
<td>a block substitution of one or more nucleotides that cause
frameshift changes in protein coding sequence</td>
</tr>
<tr class="even">
<td>stopgain</td>
<td>a nonsynonymous SNV, frameshift insertion/deletion, nonframeshift
insertion/deletion or block substitution that lead to the immediate
creation of stop codon at the variant site. For frameshift mutations,
the creation of stop codon downstream of the variant will not be counted
as “stopgain”!</td>
</tr>
<tr class="odd">
<td>stoploss</td>
<td>a nonsynonymous SNV, frameshift insertion/deletion, nonframeshift
insertion/deletion or block substitution that lead to the immediate
elimination of stop codon at the variant site</td>
</tr>
<tr class="even">
<td>nonframeshift insertion</td>
<td>an insertion of 3 or multiples of 3 nucleotides that do not cause
frameshift changes in protein coding sequence</td>
</tr>
<tr class="odd">
<td>nonframeshift deletion</td>
<td>a deletion of 3 or mutliples of 3 nucleotides that do not cause
frameshift changes in protein coding sequence</td>
</tr>
<tr class="even">
<td>nonframeshift block substitution</td>
<td>a block substitution of one or more nucleotides that do not cause
frameshift changes in protein coding sequence</td>
</tr>
<tr class="odd">
<td>nonsynonymous SNV</td>
<td>a single nucleotide change that cause an amino acid change</td>
</tr>
<tr class="even">
<td>synonymous SNV</td>
<td>a single nucleotide change that does not cause an amino acid
change</td>
</tr>
<tr class="odd">
<td>unknown</td>
<td>unknown function (due to various errors in the gene structure
definition in the database file)</td>
</tr>
</tbody>
</table>
<p>(+) see <em>Annovar</em></p>
</div>
<div class="section level4">
<h4 id="single-cell-sampling-depth">Single cell sampling depth<a class="anchor" aria-label="anchor" href="#single-cell-sampling-depth"></a>
</h4>
<p>The number of reads per SNV site represents an optimal filter to
retrieve relevant mutations. Depth at specific sites is usually not
provided in standard alignment or variant calling pipelines and it is
computationally expensive to retrieve. The user should provide the
folder with aligned data and, if not found already, the
<em>samtools</em> executable folder location to compute depth only on
sites passing the variant filters set in the previous tab.</p>
<p><img src="resources%2FDepth_tab.png" style="width:95.0%"></p>
</div>
</div>
<div class="section level3">
<h3 id="selection-of-relevant-variants">Selection of relevant variants<a class="anchor" aria-label="anchor" href="#selection-of-relevant-variants"></a>
</h3>
<p>Not all mutations are distinctive of the disease or experiment
understudied. Identifying relevant and driver variants allows to
reproduce a more significant longitudinal clonal tree. The user can
select a set of filters based on gold standards and other analyses such
as:</p>
<ol style="list-style-type: decimal">
<li>the minimum number of reads at given mutation site (field 4.1),</li>
<li>the maximum number of missing data per gene (field 4.2),</li>
<li>the minimum median depth per locus (field 4.3),</li>
<li>the minimum median depth supporting the mutation (field 4.4),</li>
<li>subset of known genes (field 4.5).</li>
</ol>
<p><img src="resources%2Ffigure4d.png" style="width:95.0%"></p>
<p>Finding the parameters to select variants is not an easy task, and
the user might not know in advance how to choose the best set of
filters. Hence, the user can press “Select variants” (<img src="resources%2Fselect_variant.png" style="width:10.0%">) to perform
the aforementioned computations on VCF and BAM files to derive all the
necessary aggregated information on the sampled cells. Afterward, the
user is presented with a interactive live preview of variants passing
the filters (including relevant parameters which can help in the
selection) while values are changed.</p>
</div>
<div class="section level3">
<h3 id="inference">Inference<a class="anchor" aria-label="anchor" href="#inference"></a>
</h3>
<p>The inferential tab allows the user to set all the parameters to
solve the boolean matrix factorization problem and estimate the model
parameters of the Bayesian model using a MCMC to maximize the
likelihood.</p>
<p>Inferential step uses the following set of parameters:</p>
<ol style="list-style-type: decimal">
<li>Learning rate (field 5.1),</li>
<li>False positive rates for each sample (field 5.2),</li>
<li>False negative rates for each sample (field 5.3),</li>
<li>Number of iterations in each MCMC search (field 5.4)</li>
<li>Number of restart for the MCMC (field 5.5),</li>
<li>Early stopping number of iterations with no growing likelihood
(field 5.6),</li>
<li>Number of parallel processes (field 5.7),</li>
<li>Random seed to recreate simulations (field 5.8),</li>
<li>Initialize the clonal tree randomly (field 5.9),</li>
<li>Marginalize the cell attachment matrix (field 5.10),</li>
<li>Keep equivalent solutions and return all of them (field 5.11),</li>
<li>Check indistinguishable event and remove them (field 5.12),</li>
<li>Estimate error rates of MCMC moves (field 5.13),</li>
<li>Show results (field 5.14).</li>
</ol>
<p><img src="resources%2FInference_tab1.png" style="width:95.0%"></p>
<p>The user must insert at least one false positive rate (field 5.2) and
one false negative rate value (field 5.3) for each sample. By double
clicking on any cell belonging to the row “add row”, the user can insert
a new set of false or positive rates for each sample. During the
inferential step, the maximization of the likelihood for each set of
rates is performed. The best results are returned.</p>
<p><img src="resources%2FInference_tab2.png" style="width:40.0%"></p>
<p>Press “Run LACE” to perform all computation and estimation steps and
visualize the results.</p>
<div class="section level4">
<h4 id="parameter-ranges">Parameter ranges<a class="anchor" aria-label="anchor" href="#parameter-ranges"></a>
</h4>
<p>To analyze new datasets, users can start from default values and then
adjust them according to the indications reported in the following
table:</p>
<table class="table">
<colgroup>
<col width="30%">
<col width="23%">
<col width="17%">
<col width="28%">
</colgroup>
<thead><tr class="header">
<th>Field Name</th>
<th>Default</th>
<th>Range</th>
<th>Variation</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Alternate counts (field 3.1)</td>
<td>2</td>
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>1</mn><mo>&lt;</mo><mi>x</mi><mo>≪</mo><mi>C</mi><mi>/</mi><mn>2</mn></mrow><annotation encoding="application/x-tex">1 &lt; x \ll C/2</annotation></semantics></math></td>
<td>Higher values select events with more supporting ALT reads and
remove PCR errors, but bias the results toward high CN/expressed
genes.</td>
</tr>
<tr class="even">
<td>MAF (field 3.2)</td>
<td>0.01</td>
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>0</mn><mo>≤</mo><mi>x</mi><mo>&lt;</mo><mn>0.5</mn></mrow><annotation encoding="application/x-tex">0 \leq x &lt; 0.5</annotation></semantics></math></td>
<td>Smaller values imply more rare events compared to the reference
population.</td>
</tr>
<tr class="odd">
<td>Variant Frequency (field 3.3)</td>
<td>0.01</td>
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>0</mn><mo>&lt;</mo><mi>x</mi><mo>≤</mo><mn>1</mn></mrow><annotation encoding="application/x-tex">0 &lt; x \leq 1</annotation></semantics></math></td>
<td>Smaller values include less common mutations among cells in the
experiment.</td>
</tr>
<tr class="even">
<td>Minimum depth (field 4.1)</td>
<td>3</td>
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>0</mn><mo>≤</mo><mi>x</mi><mo>≪</mo><mi>C</mi></mrow><annotation encoding="application/x-tex">0 \leq x \ll C</annotation></semantics></math></td>
<td>Bigger values imply more NA sites per time point.</td>
</tr>
<tr class="odd">
<td>Max missing value (field 4.2)</td>
<td>0.4</td>
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>0</mn><mo>≤</mo><mi>x</mi><mo>≤</mo><mn>1</mn></mrow><annotation encoding="application/x-tex">0\leq x\leq 1</annotation></semantics></math></td>
<td>Higher values keep mutational sites for which larger number of cells
has an NA value.</td>
</tr>
<tr class="even">
<td>Minimum median depth (field 4.3)</td>
<td>8</td>
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>1</mn><mo>≤</mo><mi>x</mi><mo>&lt;</mo><mo>∼</mo><mi>C</mi></mrow><annotation encoding="application/x-tex">1\leq x &lt; \sim C</annotation></semantics></math></td>
<td>Bigger values keep only loci with bigger median coverage among
cells.</td>
</tr>
<tr class="odd">
<td>Minimum alt median depth (field 4.4)</td>
<td>4</td>
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>1</mn><mo>≤</mo><mi>x</mi><mo>&lt;</mo><mo>∼</mo><mi>C</mi></mrow><annotation encoding="application/x-tex">1\leq x &lt; \sim C</annotation></semantics></math></td>
<td>Bigger values keep only mutations with more supporting reads in the
cell population.</td>
</tr>
<tr class="even">
<td>Learning rate (field 5.1)</td>
<td>1</td>
<td></td>
<td>Higher values permit to avoid local minima, but are less
accurate.</td>
</tr>
<tr class="odd">
<td>False positive rates (field 5.2)</td>
<td></td>
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>0</mn><mo>&lt;</mo><mi>x</mi><mo>&lt;</mo><mn>1</mn></mrow><annotation encoding="application/x-tex">0&lt;x&lt;1</annotation></semantics></math></td>
<td>Lower values imply more constraining data values.</td>
</tr>
<tr class="even">
<td>False negative rates (field 5.3)</td>
<td></td>
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>0</mn><mo>&lt;</mo><mi>x</mi><mo>&lt;</mo><mn>1</mn></mrow><annotation encoding="application/x-tex">0&lt;x&lt;1</annotation></semantics></math></td>
<td>Lower values imply more constraining data values.</td>
</tr>
<tr class="odd">
<td>MCMC iterations (field 4.4)</td>
<td>10000</td>
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>x</mi><mo>&gt;</mo><mn>1</mn></mrow><annotation encoding="application/x-tex">x&gt;1</annotation></semantics></math></td>
<td></td>
</tr>
<tr class="even">
<td>Number of restart (field 5.5)</td>
<td>50</td>
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>x</mi><mo>≥</mo><mn>1</mn></mrow><annotation encoding="application/x-tex">x\geq 1</annotation></semantics></math></td>
<td></td>
</tr>
<tr class="odd">
<td>Early stopping (field 5.6)</td>
<td>500</td>
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>x</mi><mo>&gt;</mo><mn>1</mn></mrow><annotation encoding="application/x-tex">x&gt;1</annotation></semantics></math></td>
<td></td>
</tr>
</tbody>
</table>
<p>Some of the parameter validity ranges refer to the mean mapped read
depth
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>C</mi><annotation encoding="application/x-tex">C</annotation></semantics></math>
of the samples.</p>
</div>
</div>
<div class="section level3">
<h3 id="longitudinal-display-and-outputs-interface">Longitudinal display and outputs interface<a class="anchor" aria-label="anchor" href="#longitudinal-display-and-outputs-interface"></a>
</h3>
<p>The longitudinal display tab shows the longitudinal clonal tree, the
fishplot and the clonal tree, together with the best false positive and
negative rate parameters.</p>
<p>The longitudinal clonal tree is placed on the top left side.
Continuous edges are used to represent parental relations between
clones, while dashed edges show persistence relations. The size of the
nodes is proportional to the clonal prevalence. Moving the mouse over a
node, it is possible to see the clonal prevalence value. While moving
the mouse over the mutation labels, it shows the complete set of
mutations for that clone under the longitudinal clonal tree.</p>
<p>The fishplot is on the top right side of the tab. Passing the mouse
over the ribbons, the clonal prevalences of the clones at different time
points are displayed. Similarly, with the mouse over the time tags, the
clonal prevalences of all clones at the specified time point are
visualized.</p>
<p>By clicking on a mutation label of the longitudinal clonal tree as
well as on a fishplot ribbon or a clonal tree node, the details of the
mutation characterizing the clone can be queried and retrieved from the
<em>Ensembl</em> database.</p>
<p>The bottom part of the tab is used to display the tree legend and the
best estimated set false positive/negative rates.</p>
<p><img src="resources%2Ffigure6d.png" style="width:95.0%"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Daniele Ramazzotti, Fabrizio Angaroni, Davide Maspero, Alex Graudenzi, Luca De Sano, Gianluca Ascolani.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

      </footer>
</div>






  </body>
</html>
